### DISCUSSION TITLE
detect multiple languages at once

### DISCUSSION THREAD

**Discussion by khawar-islam** [mindee/doctr#1016] [Q&A] [answered]
Most of the documents contain two languages like Arabic and English in same document. 

Can we detect multiple languages via docTR in document at once?
If not/yes, can we fine-tune previous model on new language?

**Comment by felixdittrich92** [comment, marked as answer]
Hi @khawar-islam :wave:,

currently all models are only trained on french vocab so yes you need to train your own model (especially for arabic vocab) 

modify training script as follows

```
vocab = "".join(set(VOCABS['english'] + VOCABS['arabic']))

model = recognition.__dict__[args.arch](pretrained=args.pretrained, vocab=vocab)

# example loading after training
reco_model = recognition.master(pretrained=False, vocab=vocab)
```

**Reply by khawar-islam** [reply]
Hi @felixdittrich92 

Thank you for your response. I have dataset but i want to fine tune on your recognition model because i have very limited resources and it is very hard to obtain good accuracy with small GPU.

Regards,
Khawar

**Reply by felixdittrich92** [reply]
Hi @khawar-islam ,

could you share some more information about your machine (GPU (VRAM) / RAM) and dataset size ? 
After training can the model used on GPU or only CPU ? 
Some limitations about inference time ? 

PS: GPU has nothing to do with models accuracy :) About training the hyperparameter 'batch_size' will be your best friend ;)

**Reply by khawar-islam** [reply]
The GPU is 3090 with 12 GB VRAM, 64 GB RAM and the dataset size is 194,486 images and training can be done on GPU.

**Reply by felixdittrich92** [reply]
Are you sure that it is a 3090 ? (I think all 3090 have 24 GB VRAM) 
Inference also possible on GPU ? 
dataset is mixed english arabic ?  (I think for arabic this is much to less) 
Would it be possible to increase the dataset size ?

**Reply by khawar-islam** [reply]
Sorry, its NVIDIA GeForce RTX 3080
I think inference on GPU is possible
I just give the example for arabic but i am working korean+english
Increasing dataset size is possible? How much is good?

Current condition of my dataset contains below images

**Back** 30,012 images
![두리번거리다 신세대 상당하다 코스모스 어찌 너무 뵈다 전구 내외 온라인_388](https://user-images.githubusercontent.com/11488932/184315096-12b5c86c-0bfa-4fcd-acdf-fce6b28a6573.jpg)


**Basic** 35,652 images
![두리번거리다 나흘 제과점 강조하다 신선하다 찾아내다 킬로 없이 싱싱하다 차남_783](https://user-images.githubusercontent.com/11488932/184314956-cb1adf3d-e2da-4085-8c00-e780d74bd8dd.jpg)


**Blur** 30,024 images
![두리번거리다 일시적 사회학 절 모자 보호하다 종이컵 시디 위험성 감히_200](https://user-images.githubusercontent.com/11488932/184314242-82028237-bb0a-4f27-bad1-faf8460079f1.jpg)

**Dist** 38,779 images
![두리번거리다 희생하다 자랑하다 욕하다 지겹다 버튼 그리 간판 명령 억_218](https://user-images.githubusercontent.com/11488932/184314366-a5bcf0a9-fbe7-446c-b6ba-ade40e0f9e5d.jpg)

**Skew** 60,015 images
![두리번거리다 중요시하다 모기 부르다 숨지다 경찰관 새다 흉내 차량 행사_275](https://user-images.githubusercontent.com/11488932/184314615-1f12ea3f-009a-4327-b713-b474290eb6fc.jpg)

**Reply by felixdittrich92** [reply]
Ok let me provide some overview (for korean vocab - from scratch training):

CRNN: 
- max_length: 32 chars can be recognized per image (paper default value)
- dataset: from scratch ~10M should be a good value
- with mobilenet backbone really fast on CPU and GPU
- low memory consumption

SAR:
- max_length: 30 chars can be recognized per image (paper default value)
- dataset: from scratch ~10M should be a good value
- robust for some irregular cases
- slow on CPU 
- medium memory consumption

MASTER:
- max_length: 50 chars can be recognized per image (paper default value)
- dataset: from scratch ~2M should be a good value (after toy run on 500K MJSynth is was able to recognize ~70% of my real test documents which is really powerful)
- robust for some irregular cases
- slow on CPU
- high memory consumption


The following is just a guess:
I think with your synthetic dataset you will reach the best results with MASTER arch. (12 GB should be fine with amp enabled and a batch size from 16)
If possible you should train on synthetic data and validate on real data

**Reply by khawar-islam** [reply]
@felixdittrich92 thank you for brief suggestion. I will do it and get back to you soon.

**Reply by felixdittrich92** [reply]
short note: for SAR and MASTER you need to take the main branch of this repository (both models are broken in the current pip release v0.5.1)

**Reply by khawar-islam** [reply]
@felixdittrich92 I have regenerated synthetic images dataset that contains ~3M images in five different folder like skew, blur, background and etc., Can i merge all folder into one folder than divide into train and test folder. Am i correct?

**Reply by felixdittrich92** [reply]
@khawar-islam correct. Take a look at the example i have shared with you a while ago :)

**Reply by khawar-islam** [reply]
@felixdittrich92 Yes, I have remembered that example and following same procedure as you describe previously.

**Reply by khawar-islam** [reply]
@felixdittrich92 My dataset is ready for training but still i am confused about VOCABS file. 
How i can write a VOCABS? If you can give any suggestion, I will be extremely thankful to you

`https://github.com/mindee/doctr/blob/main/doctr/datasets/vocabs.py`

**Reply by felixdittrich92** [reply]
Hi @khawar-islam 

1. you need to add korean vocab like this (you could also open an PR to add this vocab permanent)
`VOCABS['korean'] = <add all korean characters in one string> `
2. update training script (for example pytorch) https://github.com/mindee/doctr/blob/2ea3b54a4ffb86dc7586b5cd47c898544d3b2b10/references/recognition/train_pytorch.py#L185
to:  `vocab = "".join(set(VOCABS['english'] + VOCABS['korean']))  # merge vocabs and remove duplicates`
3. trigger training:
`python references/recognition/train_pytorch.py master --train_path path/to/your/train_set --val_path path/to/your/val_set  --epochs 5`

**Comment by khawar-islam** [comment]
@felixdittrich92 thank you. Sure! I will open PR if below is correct while searching on Google. Actually, I read a lot of questions and answers in your repo and there are some characters problem in Persian language thats why i am little bit worry.

![image](https://user-images.githubusercontent.com/11488932/186054352-082f7560-d274-4174-a4ad-fa5e1c9f0945.png)

`VOCABS['korean'] = "ㄱㄴㄷㄹㅁㅂㅅㅇㅈㅊㅋㅌㅍㅎㅏㅑㅓㅕㅗㅛㅜㅠㅡㅣ"`

**Reply by felixdittrich92** [reply]
Hi @khawar-islam :wave:,
Should work:
```
korean_vocab = "ㄱㄴㄷㄹㅁㅂㅅㅇㅈㅊㅋㅌㅍㅎㅏㅑㅓㅕㅗㅛㅜㅠㅡㅣ"

encoded = korean_vocab.encode('utf-8')
print(encoded)
decoded = encoded.decode('utf-8')
print(decoded)
```

```
b'\xe3\x84\xb1\xe3\x84\xb4\xe3\x84\xb7\xe3\x84\xb9\xe3\x85\x81\xe3\x85\x82\xe3\x85\x85\xe3\x85\x87\xe3\x85\x88\xe3\x85\x8a\xe3\x85\x8b\xe3\x85\x8c\xe3\x85\x8d\xe3\x85\x8e\xe3\x85\x8f\xe3\x85\x91\xe3\x85\x93\xe3\x85\x95\xe3\x85\x97\xe3\x85\x9b\xe3\x85\x9c\xe3\x85\xa0\xe3\x85\xa1\xe3\x85\xa3'
ㄱㄴㄷㄹㅁㅂㅅㅇㅈㅊㅋㅌㅍㅎㅏㅑㅓㅕㅗㅛㅜㅠㅡㅣ
```

**Reply by felixdittrich92** [reply]
Btw. if you open a PR and add korean to vocabs please don't forget to update:
https://github.com/mindee/doctr/blob/main/docs/source/modules/datasets.rst  also :)  (the table with provided vocabs at the end)

**Reply by khawar-islam** [reply]
@felixdittrich92 I did PR and update the table as well. Soon, I will make a simple tutorial on data generation and generating label file and how to do training?

**Reply by felixdittrich92** [reply]
Hi @khawar-islam :wave:, i cannot see your PR have you opened it ? https://github.com/mindee/doctr/pulls (Let me know if you need some help / how to) :+1: 
About the tutorial: We have a repository for something like this https://github.com/mindee/notebooks
If it works well feel free to add it as PR :hugs:

**Reply by khawar-islam** [reply]
@felixdittrich92 I will update you soon about PR.
